### DISCUSSION TITLE
Custom Handler example

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [NRWLDev/fastapi-problem#8] [Q&A] [answered]
In the example at https://nrwldev.github.io/fastapi-problem/handlers/, what is

`from my_module.error import CustomBaseError`

Also, in `my_custom_handler`, what is

`details=exc.debug`

**Comment by EdgyEdgemond** [comment]
`MyCustomError` is an import of any user defined exception base class, and `exc.debug` is accessing a custom attribute belonging to that custom class. 

I'll expand those docs to include an implementation example of a custom exception, i e. not a subclass of `Problem`

**Reply by EdgyEdgemond** [reply]
It could also be used when a third party library defines an exception class that you wish to handle, without catching and converting it to a `Problem` at every point it gets raised.

**Comment by EdgyEdgemond** [comment]
Documentation has been updated: https://nrwldev.github.io/fastapi-problem/handlers/

I hope this makes the custom handler use case a little clearer.

**Reply by dineshbvadhia** [reply]
Ok. I'm trying to understand how best to use fastapi-problem server-side. Client requests come to the fastapi server through an api. The error code handling the api requests is of the form:
```
from fastapi.exceptions import RequestValidationError

def verify_a(a):
	if do_a(a):
	    return a
	else:
	    raise RequestValidationError(msg_a)

def verify_b(b):
	if do_b(b):
	    return b
	else:
	    raise RequestValidationError(msg_b)

etc etc.
```
On error only the `details` (= msg_a, msg_b, etc.) change.

How is a handler for this created? Is it like the handler in the tests ?:
`https://github.com/NRWLDev/fastapi-problem/blob/59c34875570b3479acfdee08b6c3d13cccc51361/tests/handler/test_fastapi.py#L70`

**Reply by EdgyEdgemond** [reply, marked as answer]
All sublasses of Problem are automatically handled by the exception_handler added to the app using `add_exception_handler`. There should be no need for any other action to have them handled.
The specific handler for that one is here 
https://github.com/NRWLDev/fastapi-problem/blob/59c34875570b3479acfdee08b6c3d13cccc51361/src/fastapi_problem/handler/fastapi.py#L23

**Reply by EdgyEdgemond** [reply]
This *should* be close to a minimal working example, away from laptop currently.

```
from typing import Union

from fastapi import FastAPI
from fastapi.exceptions import RequestValidationError
from fastapi_problem.handler.fastapi import add_exception_handler

app = FastAPI()
add_exception_handler(app)

@app.get("/")
def read_root():
    return {"Hello": "World"}


@app.get("/items/{item_id}")
def read_item(item_id: int, q: Union[str, None] = None):
    raise RequestValidationError("message")
```

**Reply by dineshbvadhia** [reply]
This works - thank-you - and is almost the same as I had with https://github.com/vapor-ware/fastapi-rfc7807. 

The next question is how to handle the other types of errors. With fastapi-rfc7807, each error was transformed into a Problem dictionary. I extracted the status-code and then processed for each category ie.
```
if status_code >= 500:
	do_this()
elif status_code >= 400:
	do_that()
elif status_code >= 300:
	do_other()
etc etc
```
This is a very ugly solution :(

**Reply by dineshbvadhia** [reply]
Ah, just looked at the https://github.com/NRWLDev/fastapi-problem/blob/main/src/fastapi_problem/handler/fastapi.py code and as you say "All sublasses of Problem are automatically handled by the exception_handler added to the app using `add_exception_handler`. There should be no need for any other action to have them handled."

Ok, I get it now.
